File: src/agents/alerter/src/farmos/assets.js
'use strict';

// Phase 40 fungi asset creation primitive. Option A hybrid shape
// (2026-05-14): asset--fungi requires fungi_type (strain code term, e.g.
// SHI/SH2/...) AND fungi_xing (structural classifier: block | fruit).
// Pre-inoc substrates are NOT fungi assets -- they live in the material
// bundle or as pasteurization logs. See
// .planning/notes/2026-05-14-reply-from-farmos-fungi-schema.md.
//
// Module-level LRU cache for asset name -> id resolution.

const qr = require('./qr');
const fungiTypeCache = require('./fungi-type-cache');
const fungiXingCache = require('./fungi-xing-cache');
const { mergeAssetFields, IdentityMutationError } = require('./merge');

// Phase 51 UPSERT-05: marker string in notes.value identifies hand-stubbed
// ancestors awaiting 2025-paper-scan backfill. See
// .planning/notes/2026-05-24-prod-write-receipt.md (4 stubs in prod farmOS).
const STUB_BACKFILL_MARKER = 'STUB - awaits 2025-paper-scan backfill';

const NAME_CACHE = new Map(); // name -> assetId; capped at 32
const NAME_CACHE_MAX = 32;

// NAME_CACHE survives PATCH without invalidation because UPSERT-03's
// IdentityMutationError on name change makes (name -> id) stable. If a future
// feature adds rename support, the cache MUST be invalidated in upsertFungiAsset.

function _cacheGet(name) {
  if (!NAME_CACHE.has(name)) return undefined;
  const v = NAME_CACHE.get(name);
  NAME_CACHE.delete(name);
  NAME_CACHE.set(name, v);
  return v;
}

function _cacheSet(name, id) {
  if (NAME_CACHE.has(name)) NAME_CACHE.delete(name);
  NAME_CACHE.set(name, id);
  while (NAME_CACHE.size > NAME_CACHE_MAX) {
    const k = NAME_CACHE.keys().next().value;
    NAME_CACHE.delete(k);
  }
}

function _clearCache() { NAME_CACHE.clear(); }

async function findAssetByName(client, name) {
  const cached = _cacheGet(name);
  if (cached) return { found: true, assetId: cached, cached: true };
  const enc = encodeURIComponent(name);
  const r = await client.get(`/api/asset/fungi?filter[name][value]=${enc}`);
  if (!r.ok) return { found: false, error: 'http_' + (r.status || 'network') };
  const arr = r.body && r.body.data;
  if (Array.isArray(arr) && arr.length > 0) {
    const id = arr[0].id;
    _cacheSet(name, id);
    return { found: true, assetId: id };
  }
  return { found: false };
}

// Internal payload builder shared by createFungiAsset and upsertFungiAsset.
// Returns either {ok:true, payload, attributes, relationships} where
// attributes/relationships are the canonical shape mergeAssetFields expects,
// or {ok:false, reason} on a taxonomy resolution failure.
async function _buildAssetBody(client, opts) {
  const {
    name,
    parentIds = [],
    fungiTypeName = null,
    fungiXingName = null,
    draftId,
    qrCodes = [],
    notes = null,
    allowNoFungiType = false,
    // Phase 54 Cycle-1: backfill (santi source-of-truth) opts in to minting an
    // unknown strain's fungi_type term instead of failing fungi_type_not_found.
    // Live capture leaves this false so extraction typos never pollute the taxonomy.
    createMissingFungiType = false,
  } = opts;
  if (!allowNoFungiType && !fungiTypeName) return { ok: false, reason: 'missing_fungi_type_name' };
  if (!fungiXingName) return { ok: false, reason: 'missing_fungi_xing_name' };
  let ft = null;
  if (fungiTypeName) {
    ft = await fungiTypeCache.ensureFungiTypeUuid(client, fungiTypeName, { create: createMissingFungiType });
    if (!ft.ok) return { ok: false, reason: ft.reason, fungiTypeName };
  }
  const fx = await fungiXingCache.getFungiXingUuid(client, fungiXingName);
  if (!fx.ok) return { ok: false, reason: fx.reason, fungiXingName };
  const noteTrailer = (notes ? notes + '\n' : '') + 'mushy:draft:' + draftId;
  const attributes = {
    name,
    status: 'active',
    notes: { value: noteTrailer, format: 'plain_text' },
  };
  const relationships = {
    fungi_xing: { data: [{ type: 'taxonomy_term--fungi_xing', id: fx.uuid }] },
  };
  if (ft && ft.uuid) {
    relationships.fungi_type = { data: [{ type: 'taxonomy_term--fungi_type', id: ft.uuid }] };
  }
  if (parentIds.length > 0) {
    relationships.parent = { data: parentIds.map((id) => ({ type: 'asset--fungi', id })) };
  }
  const payload = {
    data: {
      type: 'asset--fungi',
      attributes,
    },
  };
  payload.data.relationships = relationships;
  if (qrCodes.length > 0) {
    qr.bindQrOnCreate(payload, qrCodes);
    // Mirror the id_tag onto attributes so the merge layer treats it as part
    // of the canonical attributes view as well (payload.data.attributes is
    // the same object reference as `attributes`, so this is already done).
  }
  return { ok: true, payload, attributes, relationships };
}

async function createFungiAsset(client, opts) {
  const built = await _buildAssetBody(client, opts);
  if (!built.ok) return built;
  const r = await client.post('/api/asset/fungi', built.payload);
  if (!r.ok) {
    return { ok: false, reason: 'http_' + (r.status || 'network'), http_status: r.status };
  }
  const assetId = r.body && r.body.data && r.body.data.id;
  if (!assetId) {
    return { ok: false, reason: 'no_asset_id_in_response' };
  }
  _cacheSet(opts.name, assetId);
  return { ok: true, assetId, qrBindings: [], http_status: r.status };
}

async function resolveOrCreateAsset(client, opts) {
  const lookup = await findAssetByName(client, opts.name);
  if (lookup.found) return { ok: true, assetId: lookup.assetId, reused: true };
  return createFungiAsset(client, opts);
}

// Phase 51 UPSERT-05: stub-detection predicate. Pure.
function isStubAsset(asset) {
  if (!asset || !asset.attributes || !asset.attributes.notes) return false;
  const value = asset.attributes.notes.value;
  return typeof value === 'string' && value.includes(STUB_BACKFILL_MARKER);
}

// Helper: structural compare of merged vs existing on the dimensions we touch.
// JSON.stringify on the attributes+relationships subset is sufficient because
// mergeAssetFields produces canonical (set-union, dedup) outputs.
function _isMergeNoop(existing, merged) {
  // Compare attributes and relationships structurally, but normalize the
  // notes object to its value-only projection: createFungiAsset always
  // emits `{value, format:'plain_text'}` whereas farmOS GET responses may
  // omit `format` -- a missing-format difference is not a real change.
  function _normAttrs(a) {
    const out = Object.assign({}, a || {});
    if (out.notes && typeof out.notes === 'object') {
      out.notes = { value: out.notes.value };
    }
    // drupal_internal__revision_id is server-side metadata; never part of
    // anything we'd PATCH so exclude from compare.
    delete out.drupal_internal__revision_id;
    return out;
  }
  const ea = _normAttrs(existing && existing.attributes);
  const er = (existing && existing.relationships) || {};
  const ma = _normAttrs(merged && merged.attributes);
  const mr = (merged && merged.relationships) || {};
  return JSON.stringify(ea) === JSON.stringify(ma)
      && JSON.stringify(er) === JSON.stringify(mr);
}

// Phase 51 UPSERT-01/04/05: lookup-merge-or-create primitive.
//   - miss  -> delegate to createFungiAsset (outcome='created')
//   - hit, mergeable, non-empty diff -> PATCH with merged body (outcome='patched')
//   - hit, mergeable, zero diff      -> no PATCH (outcome='noop')
//   - hit, scalar conflict           -> no PATCH (outcome='noop', conflicts populated)
//   - hit, identity mutation         -> structured {ok:false, reason:'identity_mutation'}
//   - hit, soft revision_id moved    -> retry merge ONCE; if still moved,
//                                       outcome='noop', reason='concurrency_loss'
//
// Soft compare degrades to etag_source='absent' (and skips If-Match header)
// when the GET response has no drupal_internal__revision_id. See
// 51-RESEARCH.md §3 / Pitfall 2 — farmOS doesn't honor If-Match, the soft
// compare is a best-effort guard, not a true concurrency primitive.
async function upsertFungiAsset(client, opts) {
  const lookup = await findAssetByName(client, opts.name);

  // Miss path -> POST via createFungiAsset, wrap return.
  if (!lookup.found) {
    const r = await createFungiAsset(client, opts);
    if (!r.ok) {
      return { ok: false, reason: r.reason, http_status: r.http_status || null, conflicts: [], etag_source: null };
    }
    return {
      ok: true,
      assetId: r.assetId,
      outcome: 'created',
      conflicts: [],
      etag_source: null,
      http_status: r.http_status,
    };
  }

  // Hit path -> GET existing, build incoming, merge, decide PATCH-or-noop.
  const built = await _buildAssetBody(client, opts);
  if (!built.ok) {
    return { ok: false, reason: built.reason, http_status: null, conflicts: [], etag_source: null };
  }
  // Normalize incoming relationships to the singleton-`data` shape that
  // merge.js's SCALAR_REL_FIELDS expects. createFungiAsset POSTs
  // fungi_type/fungi_xing as `{data: [{...}]}` (array form) which is what
  // farmOS accepts on create; existing assets on GET come back as
  // `{data: {...}}` (singleton). For mergeAssetFields to compare like-with-
  // like, collapse the array form to singleton on the merge input.
  const incomingRelationships = Object.assign({}, built.relationships);
  for (const f of ['fungi_type', 'fungi_xing']) {
    const rel = incomingRelationships[f];
    if (rel && Array.isArray(rel.data)) {
      incomingRelationships[f] = { data: rel.data[0] || null };
    }
  }
  const incoming = { type: 'asset--fungi', attributes: built.attributes, relationships: incomingRelationships };

  async function _getExisting() {
    const r = await client.get('/api/asset/fungi/' + lookup.assetId);
    if (!r.ok) return null;
    return r.body && r.body.data;
  }

  // One attempt of merge cycle: GET -> merge -> (if non-noop, non-conflict)
  // -> re-GET to soft-compare revision_id -> PATCH if revision stable.
  // Returns one of: {kind:'patched', result}, {kind:'noop', result},
  // {kind:'conflict', result}, {kind:'identity', result}, {kind:'race'} (needs retry).
  async function _attempt() {
    const existing = await _getExisting();
    if (!existing) {
      return { kind: 'noop', result: { ok: false, reason: 'lookup_missing_after_find', http_status: null, conflicts: [], etag_source: null } };
    }
    const preMergeRevisionId = existing.attributes && existing.attributes.drupal_internal__revision_id != null
      ? existing.attributes.drupal_internal__revision_id
      : null;
    const etag_source = preMergeRevisionId != null ? 'soft_compare' : 'absent';

    let merged;
    let conflicts;
    try {
      const out = mergeAssetFields(existing, incoming);
      merged = out.merged;
      conflicts = out.conflicts;
    } catch (e) {
      if (e instanceof IdentityMutationError) {
        return { kind: 'identity', result: { ok: false, reason: 'identity_mutation', http_status: null, conflicts: [], etag_source: null } };
      }
      throw e;
    }

    if (conflicts.length > 0) {
      return { kind: 'conflict', result: { ok: true, assetId: lookup.assetId, outcome: 'noop', conflicts, etag_source, http_status: null } };
    }

    if (_isMergeNoop(existing, merged)) {
      return { kind: 'noop', result: { ok: true, assetId: lookup.assetId, outcome: 'noop', conflicts: [], etag_source, http_status: null } };
    }

    // Soft-compare guard: re-GET, compare revision_id with preMergeRevisionId.
    if (preMergeRevisionId != null) {
      const reGot = await _getExisting();
      const currentRevisionId = reGot && reGot.attributes && reGot.attributes.drupal_internal__revision_id != null
        ? reGot.attributes.drupal_internal__revision_id
        : null;
      if (currentRevisionId !== preMergeRevisionId) {
        return { kind: 'race' };
      }
    }

    // PATCH path.
    const headers = preMergeRevisionId != null ? { 'If-Match': String(preMergeRevisionId) } : undefined;
    const patchBody = {
      data: {
        type: 'asset--fungi',
        id: lookup.assetId,
        attributes: merged.attributes,
        relationships: merged.relationships,
      },
    };
    const pr = await client.patch('/api/asset/fungi/' + lookup.assetId, patchBody, headers ? { headers } : undefined);
    if (!pr.ok) {
      return { kind: 'patched', result: { ok: false, reason: 'http_' + (pr.status || 'network'), http_status: pr.status, conflicts: [], etag_source } };
    }
    return { kind: 'patched', result: { ok: true, assetId: lookup.assetId, outcome: 'patched', conflicts: [], etag_source, http_status: pr.status } };
  }

  const first = await _attempt();
  if (first.kind !== 'race') return first.result;
  // Retry budget = 1.
  const second = await _attempt();
  if (second.kind !== 'race') return second.result;
  // Still racing -> concurrency_loss.
  return {
    ok: true,
    assetId: lookup.assetId,
    outcome: 'noop',
    reason: 'concurrency_loss',
    conflicts: [],
    etag_source: 'soft_compare',
    http_status: null,
  };
}

// Phase 48 Plan 02: best-effort orphan cleanup after partial commit failure
// in commit-seeding-session. Caller treats non-ok as audit-log-and-continue;
// this primitive never throws.
async function deleteFungiAsset(client, assetId) {
  if (!assetId) return { ok: false, reason: 'missing_asset_id' };
  if (typeof client.delete !== 'function') {
    return { ok: false, reason: 'client_delete_unavailable' };
  }
  const r = await client.delete('/api/asset/fungi/' + assetId);
  if (!r.ok) return { ok: false, reason: 'http_' + (r.status || 'network'), http_status: r.status };
  // Invalidate the name cache for any entry that points at this id so a
  // subsequent re-commit attempt does not return a stale UUID for a deleted
  // asset. Linear scan is fine (cache is capped at 32).
  for (const [name, id] of NAME_CACHE.entries()) {
    if (id === assetId) NAME_CACHE.delete(name);
  }
  return { ok: true, http_status: r.status };
}

module.exports = {
  findAssetByName,
  createFungiAsset,
  resolveOrCreateAsset,
  deleteFungiAsset,
  _clearCache,
  upsertFungiAsset,
  isStubAsset,
  STUB_BACKFILL_MARKER,
  __test_isMergeNoop: _isMergeNoop,
};

File: src/farm-agent/farm_agent/farmos/assets.py
"""
farm_agent/farmos/assets.py -- Fungi asset upsert primitive (Phase 62-07).

Faithful Python port of src/agents/alerter/src/farmos/assets.js (Phase 40/51).

JS camelCase -> Python snake_case mapping for opts parameters:
  name                   -> name
  parentIds              -> parent_ids
  fungiTypeName          -> fungi_type_name
  fungiXingName          -> fungi_xing_name
  draftId                -> draft_id
  qrCodes                -> qr_codes
  notes                  -> notes
  allowNoFungiType       -> allow_no_fungi_type
  createMissingFungiType -> create_missing_fungi_type

Return keys are snake_case throughout (asset_id, http_status, etag_source, etc.).

Phase 40 fungi asset creation primitive. Option A hybrid shape (2026-05-14):
asset--fungi requires fungi_type (strain code term, e.g. SHI/SH2/...) AND
fungi_xing (structural classifier: block | fruit). Pre-inoc substrates are NOT
fungi assets -- they live in the material bundle or as pasteurization logs.

Module-level LRU cache for asset name -> id resolution (cap-32 OrderedDict).
NAME_CACHE survives PATCH without invalidation because UPSERT-03's
IdentityMutationError on name change makes (name -> id) stable.

ASCII-only source. No em-dashes.
"""

from __future__ import annotations

import json
import urllib.parse
from collections import OrderedDict

from farm_agent.farmos import fungi_type_cache, fungi_xing_cache, qr
from farm_agent.farmos.merge import IdentityMutationError, merge_asset_fields

# Phase 51 UPSERT-05: marker string in notes.value identifies hand-stubbed
# ancestors awaiting 2025-paper-scan backfill.
# See .planning/notes/2026-05-24-prod-write-receipt.md (4 stubs in prod farmOS).
STUB_BACKFILL_MARKER = "STUB - awaits 2025-paper-scan backfill"

_NAME_CACHE: OrderedDict[str, str] = OrderedDict()  # name -> asset_id; capped at 32
_NAME_CACHE_MAX = 32


def _cache_get(name: str) -> str | None:
    """LRU get: move-to-end on hit (mirror JS Map delete+set pattern)."""
    if name not in _NAME_CACHE:
        return None
    _NAME_CACHE.move_to_end(name)
    return _NAME_CACHE[name]


def _cache_set(name: str, asset_id: str) -> None:
    """LRU set: move-to-end, evict-oldest on overflow."""
    if name in _NAME_CACHE:
        _NAME_CACHE.move_to_end(name)
    _NAME_CACHE[name] = asset_id
    while len(_NAME_CACHE) > _NAME_CACHE_MAX:
        _NAME_CACHE.popitem(last=False)  # evict oldest (FIFO)


def _clear_cache() -> None:
    """Clear the name cache (test isolation)."""
    _NAME_CACHE.clear()


def is_stub_asset(asset: dict | None) -> bool:
    """Pure predicate: True when asset notes.value contains the STUB backfill marker.

    Port of isStubAsset() from assets.js lines 143-148.
    """
    if not asset:
        return False
    attrs = asset.get("attributes") or {}
    notes = attrs.get("notes")
    if not notes:
        return False
    value = notes.get("value")
    return isinstance(value, str) and STUB_BACKFILL_MARKER in value


async def find_asset_by_name(client: dict, name: str) -> dict:
    """Resolve name -> asset_id via name-filter GET on /api/asset/fungi.

    Port of findAssetByName() from assets.js lines 48-61.
    Uses the name-value filter query parameter (D-05 name-based stable identity).
    LRU-cached (cap-32). Transport failure is NOT a miss.

    Returns:
      {"found": True,  "asset_id": str}                   -- resolved fresh
      {"found": True,  "asset_id": str, "cached": True}   -- LRU cache hit
      {"found": False}                                     -- not found (empty data)
      {"found": False, "error": "http_<status|network>"}  -- transport failure
    """
    cached = _cache_get(name)
    if cached:
        return {"found": True, "asset_id": cached, "cached": True}
    enc = urllib.parse.quote(name, safe="")
    r = await client["get"](f"/api/asset/fungi?filter[name][value]={enc}")
    if not r["ok"]:
        return {
            "found": False,
            "error": "http_" + (str(r.get("status")) if r.get("status") else "network"),
        }
    arr = (r.get("body") or {}).get("data")
    if isinstance(arr, list) and arr:
        asset_id = arr[0]["id"]
        _cache_set(name, asset_id)
        return {"found": True, "asset_id": asset_id}
    return {"found": False}


# MUSHY-133: every lookup in this module was hardcoded to the fungi bundle, so
# the agent could not see the other 34 assets on the farm. Kimba the farm dog is
# asset--animal id 1; an observation naming her failed as though she did not
# exist, and ref_check offered to "create" her.
#
# Fungi first so the common path costs one request and keeps the name cache.
# Creation stays fungi-only on purpose: this list is for RESOLUTION.
RESOLVABLE_BUNDLES = ("fungi", "animal", "plant", "structure", "group", "land", "equipment", "water")


async def find_asset_any_bundle(client: dict, name: str) -> dict:
    """Resolve name -> (asset_id, bundle) across every asset bundle.

    Deliberately NOT a widening of find_asset_by_name. That one feeds
    upsert_fungi_asset's create-or-patch decision, and if it started returning
    an animal that happens to share a block's name, the upsert would PATCH the
    animal as a fungi asset. Resolution and creation need different scopes.

    Only the fungi step touches the LRU name cache, for the same reason: a
    cached non-fungi id would leak into the upsert path through the cache.

    A transport failure on ANY bundle is returned as an error rather than a
    miss, matching find_asset_by_name. "I could not look" is not "not there".

    Returns:
      {"found": True, "asset_id": str, "bundle": str}
      {"found": False}
      {"found": False, "error": "http_<status|network>"}
    """
    fungi = await find_asset_by_name(client, name)
    if fungi.get("found"):
        return {"found": True, "asset_id": fungi["asset_id"], "bundle": "fungi"}
    if fungi.get("error"):
        return fungi

    enc = urllib.parse.quote(name, safe="")
    for bundle in RESOLVABLE_BUNDLES:
        if bundle == "fungi":
            continue
        r = await client["get"](f"/api/asset/{bundle}?filter[name][value]={enc}")
        if not r["ok"]:
            status = r.get("status")
            # A bundle this farmOS does not expose is a 404 on the collection,
            # which is a fact about the schema, not a failed lookup. Keep going.
            if status == 404:
                continue
            return {
                "found": False,
                "error": "http_" + (str(status) if status else "network"),
            }
        arr = (r.get("body") or {}).get("data")
        if isinstance(arr, list) and arr:
            return {"found": True, "asset_id": arr[0]["id"], "bundle": bundle}
    return {"found": False}


async def _build_asset_body(client: dict, opts: dict) -> dict:
    """Internal payload builder shared by create_fungi_asset and upsert_fungi_asset.

    Port of _buildAssetBody() from assets.js lines 67-120.

    Returns {"ok": True, "payload": dict, "attributes": dict, "relationships": dict}
    or {"ok": False, "reason": str, ...} on taxonomy resolution failure.
    """
    name = opts.get("name")
    parent_ids = opts.get("parent_ids") or []
    fungi_type_name = opts.get("fungi_type_name") or None
    fungi_xing_name = opts.get("fungi_xing_name") or None
    draft_id = opts.get("draft_id")
    qr_codes = opts.get("qr_codes") or []
    notes = opts.get("notes") or None
    allow_no_fungi_type = opts.get("allow_no_fungi_type", False)
    create_missing_fungi_type = opts.get("create_missing_fungi_type", False)

    if not allow_no_fungi_type and not fungi_type_name:
        return {"ok": False, "reason": "missing_fungi_type_name"}
    if not fungi_xing_name:
        return {"ok": False, "reason": "missing_fungi_xing_name"}

    ft = None
    if fungi_type_name:
        ft = await fungi_type_cache.ensure_fungi_type_uuid(
            client, fungi_type_name, create=create_missing_fungi_type
        )
        if not ft["ok"]:
            return {"ok": False, "reason": ft["reason"], "fungi_type_name": fungi_type_name}

    fx = await fungi_xing_cache.get_fungi_xing_uuid(client, fungi_xing_name)
    if not fx["ok"]:
        return {"ok": False, "reason": fx["reason"], "fungi_xing_name": fungi_xing_name}

    note_trailer = (notes + "\n" if notes else "") + "mushy:draft:" + str(draft_id)
    attributes: dict = {
        "name": name,
        "status": "active",
        "notes": {"value": note_trailer, "format": "plain_text"},
    }
    relationships: dict = {
        "fungi_xing": {"data": [{"type": "taxonomy_term--fungi_xing", "id": fx["uuid"]}]},
    }
    if ft and ft.get("uuid"):
        relationships["fungi_type"] = {
            "data": [{"type": "taxonomy_term--fungi_type", "id": ft["uuid"]}]
        }
    if parent_ids:
        relationships["parent"] = {
            "data": [{"type": "asset--fungi", "id": pid} for pid in parent_ids]
        }

    payload: dict = {
        "data": {
            "type": "asset--fungi",
            "attributes": attributes,
        },
    }
    payload["data"]["relationships"] = relationships
    if qr_codes:
        qr.bind_qr_on_create(payload, qr_codes)

    return {"ok": True, "payload": payload, "attributes": attributes, "relationships": relationships}


async def create_fungi_asset(client: dict, opts: dict) -> dict:
    """POST /api/asset/fungi to create a new fungi asset.

    Port of createFungiAsset() from assets.js lines 122-135.

    Returns {"ok": True, "asset_id": str, "qr_bindings": [], "http_status": int}
    or {"ok": False, "reason": str, "http_status": int?}.
    """
    built = await _build_asset_body(client, opts)
    if not built["ok"]:
        return built
    r = await client["post"]("/api/asset/fungi", built["payload"])
    if not r["ok"]:
        return {
            "ok": False,
            "reason": "http_" + (str(r.get("status")) if r.get("status") else "network"),
            "http_status": r.get("status"),
        }
    asset_id = ((r.get("body") or {}).get("data") or {}).get("id")
    if not asset_id:
        return {"ok": False, "reason": "no_asset_id_in_response"}
    _cache_set(opts["name"], asset_id)
    return {"ok": True, "asset_id": asset_id, "qr_bindings": [], "http_status": r.get("status")}


async def resolve_or_create_asset(client: dict, opts: dict) -> dict:
    """Find by name; if not found, create.

    Port of resolveOrCreateAsset() from assets.js lines 137-141.
    """
    lookup = await find_asset_by_name(client, opts["name"])
    if lookup["found"]:
        return {"ok": True, "asset_id": lookup["asset_id"], "reused": True}
    return await create_fungi_asset(client, opts)


def _is_merge_noop(existing: dict | None, merged: dict | None) -> bool:
    """Structural compare of merged vs existing on attributes + relationships.

    Port of _isMergeNoop() from assets.js lines 153-174.
    Normalizes notes to value-only projection and drops drupal_internal__revision_id
    before comparing (both are server-side metadata irrelevant to PATCH decisions).
    Uses sort_keys=True to normalize key insertion order.
    """
    def _norm_attrs(a: dict | None) -> dict:
        out = dict(a or {})
        if out.get("notes") and isinstance(out["notes"], dict):
            out["notes"] = {"value": out["notes"].get("value")}
        out.pop("drupal_internal__revision_id", None)
        return out

    ea = _norm_attrs((existing or {}).get("attributes"))
    er = (existing or {}).get("relationships") or {}
    ma = _norm_attrs((merged or {}).get("attributes"))
    mr = (merged or {}).get("relationships") or {}
    return (
        json.dumps(ea, sort_keys=True) == json.dumps(ma, sort_keys=True)
        and json.dumps(er, sort_keys=True) == json.dumps(mr, sort_keys=True)
    )


async def upsert_fungi_asset(client: dict, opts: dict) -> dict:
    """Lookup-merge-or-create primitive for asset--fungi.

    Port of upsertFungiAsset() from assets.js lines 189-312 (Phase 51 UPSERT-01/04/05).

    Outcomes:
      miss   -> POST via create_fungi_asset -> outcome='created'
      hit, no structural diff -> no PATCH -> outcome='noop'
      hit, diff               -> PATCH merged body -> outcome='patched'
      hit, scalar conflict    -> no PATCH -> outcome='noop', conflicts non-empty
      identity mutation       -> ok=False, reason='identity_mutation'
      concurrency race (soft revision moved) -> retry once; still racing ->
        outcome='noop', reason='concurrency_loss'

    Soft compare degrades to etag_source='absent' when the GET response has no
    drupal_internal__revision_id (no If-Match header sent in that case).
    """
    lookup = await find_asset_by_name(client, opts["name"])

    # Miss path -> POST via create_fungi_asset.
    if not lookup["found"]:
        r = await create_fungi_asset(client, opts)
        if not r["ok"]:
            return {
                "ok": False,
                "reason": r.get("reason"),
                "http_status": r.get("http_status"),
                "conflicts": [],
                "etag_source": None,
            }
        return {
            "ok": True,
            "asset_id": r["asset_id"],
            "outcome": "created",
            "conflicts": [],
            "etag_source": None,
            "http_status": r.get("http_status"),
        }

    # Hit path -> GET existing, build incoming, merge, decide PATCH-or-noop.
    built = await _build_asset_body(client, opts)
    if not built["ok"]:
        return {
            "ok": False,
            "reason": built.get("reason"),
            "http_status": None,
            "conflicts": [],
            "etag_source": None,
        }

    # Normalize incoming relationships to singleton-data shape that merge.py's
    # SCALAR_REL_FIELDS expects. create_fungi_asset POSTs fungi_type/fungi_xing as
    # `{data: [{...}]}` (array form accepted on create); existing assets on GET come
    # back as `{data: {...}}` (singleton). Collapse for like-with-like comparison.
    incoming_relationships = dict(built["relationships"])
    for field in ("fungi_type", "fungi_xing"):
        rel = incoming_relationships.get(field)
        if rel and isinstance(rel.get("data"), list):
            incoming_relationships[field] = {
                "data": rel["data"][0] if rel["data"] else None
            }

    incoming = {
        "type": "asset--fungi",
        "attributes": built["attributes"],
        "relationships": incoming_relationships,
    }

    async def _get_existing() -> dict | None:
        r = await client["get"]("/api/asset/fungi/" + lookup["asset_id"])
        if not r["ok"]:
            return None
        return (r.get("body") or {}).get("data")

    # One merge-cycle attempt: GET existing -> merge -> soft-revision re-GET -> PATCH or noop.
    # Returns {"kind": str, "result": dict?}. kind="race" means caller should retry.
    async def _attempt() -> dict:
        existing = await _get_existing()
        if not existing:
            return {
                "kind": "noop",
                "result": {
                    "ok": False,
                    "reason": "lookup_missing_after_find",
                    "http_status": None,
                    "conflicts": [],
                    "etag_source": None,
                },
            }

        attrs = existing.get("attributes") or {}
        pre_merge_rev = attrs.get("drupal_internal__revision_id")
        etag_source = "soft_compare" if pre_merge_rev is not None else "absent"

        try:
            out = merge_asset_fields(existing, incoming)
            merged = out["merged"]
            conflicts = out["conflicts"]
        except IdentityMutationError:
            return {
                "kind": "identity",
                "result": {
                    "ok": False,
                    "reason": "identity_mutation",
                    "http_status": None,
                    "conflicts": [],
                    "etag_source": None,
                },
            }

        if conflicts:
            return {
                "kind": "conflict",
                "result": {
                    "ok": True,
                    "asset_id": lookup["asset_id"],
                    "outcome": "noop",
                    "conflicts": conflicts,
                    "etag_source": etag_source,
                    "http_status": None,
                },
            }

        if _is_merge_noop(existing, merged):
            return {
                "kind": "noop",
                "result": {
                    "ok": True,
                    "asset_id": lookup["asset_id"],
                    "outcome": "noop",
                    "conflicts": [],
                    "etag_source": etag_source,
                    "http_status": None,
                },
            }

        # Soft-compare guard: re-GET to verify revision_id is still stable.
        if pre_merge_rev is not None:
            re_got = await _get_existing()
            current_rev = (re_got and (re_got.get("attributes") or {}).get("drupal_internal__revision_id"))
            if current_rev != pre_merge_rev:
                return {"kind": "race"}

        # PATCH path.
        headers = {"If-Match": str(pre_merge_rev)} if pre_merge_rev is not None else None
        patch_body = {
            "data": {
                "type": "asset--fungi",
                "id": lookup["asset_id"],
                "attributes": merged.get("attributes"),
                "relationships": merged.get("relationships"),
            },
        }
        pr = await client["patch"](
            "/api/asset/fungi/" + lookup["asset_id"],
            patch_body,
            {"headers": headers} if headers else None,
        )
        if not pr["ok"]:
            return {
                "kind": "patched",
                "result": {
                    "ok": False,
                    "reason": "http_" + (str(pr.get("status")) if pr.get("status") else "network"),
                    "http_status": pr.get("status"),
                    "conflicts": [],
                    "etag_source": etag_source,
                },
            }
        return {
            "kind": "patched",
            "result": {
                "ok": True,
                "asset_id": lookup["asset_id"],
                "outcome": "patched",
                "conflicts": [],
                "etag_source": etag_source,
                "http_status": pr.get("status"),
            },
        }

    first = await _attempt()
    if first["kind"] != "race":
        return first["result"]
    # Retry budget = 1.
    second = await _attempt()
    if second["kind"] != "race":
        return second["result"]
    # Still racing -> concurrency_loss.
    return {
        "ok": True,
        "asset_id": lookup["asset_id"],
        "outcome": "noop",
        "reason": "concurrency_loss",
        "conflicts": [],
        "etag_source": "soft_compare",
        "http_status": None,
    }


async def delete_fungi_asset(client: dict, asset_id: str) -> dict:
    """Best-effort DELETE /api/asset/fungi/<id> + name-cache invalidation.

    Port of deleteFungiAsset() from assets.js lines 317-331.
    Never raises; caller treats non-ok as audit-log-and-continue.
    Linear scan of the LRU cache (capped at 32) to invalidate stale name->id entries.
    """
    if not asset_id:
        return {"ok": False, "reason": "missing_asset_id"}
    if not callable(client.get("delete")):
        return {"ok": False, "reason": "client_delete_unavailable"}
    r = await client["delete"]("/api/asset/fungi/" + asset_id)
    if not r["ok"]:
        return {
            "ok": False,
            "reason": "http_" + (str(r.get("status")) if r.get("status") else "network"),
            "http_status": r.get("status"),
        }
    # Invalidate any cache entry pointing at this asset_id.
    to_remove = [name for name, aid in _NAME_CACHE.items() if aid == asset_id]
    for name in to_remove:
        del _NAME_CACHE[name]
    return {"ok": True, "http_status": r.get("status")}
